py/dml/template.py

```diff
@@ -144,6 +162,49 @@ def traits(self):
             return Set().union(
                 *[sup.traits() for (_, sup) in self.spec.templates])
 
+    def get_potential_method_impl_details(self, method_name):
```

Comment:
Need docstring/comments, in particular the contract for the return value.

If it has no viable use outside tqmic, then possibly better expressed as a top-level function.

Reply:
Oh right, I started writing a docstring, but it ended up needing so many revisions that I said "y'know what, I'll save the docstring for when I have actually figured this out" and then I forgot about it

Reply:
> If it has no viable use outside tqmic, then possibly better expressed as a top-level function.
It is not _hyper_-specific to TQMIC, but it could be that it is too specific. What this method essentially does (through repeated calls of it) is provide you with a tree of the current and ancestor templates, filtered by which may provide the highest-rank implementation of the named method, ("may" because the definition could be conditionally eliminated). It's possible that this functionality can be generalized to param definitions as well.
That is interesting if you need to get an idea of the possible implementations that a specific template provides, which can possibly be interesting for static analysis? Or if you need information about all method implementations provided by templates to to an object, and not just the final implementations the object has, which is what TQMIC uses it for.

Reply:
> It's possible that this functionality can be generalized to param definitions as well.

I bet we will get a feature request for TQPIC pretty soon, but I have a feeling that we should handle that with a separate modified copy of the code path for methods (methods and params are too dissimilar for that kind of abstraction).

Reply:
`functools.cache` makes it feel natural to keep it a method (although top-level would still work of course).

Reply:
This is for you to decide; I'd rather keep it a method. Moving it to top-level would be due to your insistence.

Reply:
The purpose of my previous comment was to lift my insistence. Method is better.